Addition of json_source module

evgepab opened the pull request:
The `json_source` module contains:

* `JSONElementsCursor`, which is identical to both how `CrossrefElementsCursor` and `DataciteElemenrsCursors` were and
* `RecordsCursor`, which contains the methods from the `WorksCursor` 's classes from the Crossref and the DataCite modules, with an exception of the `current_row_value` method which called for an abstract implementation

dspinellis commented on src/alexandria3k/data_sources/crossref.py:
```diff
@@ -18,18 +18,15 @@
 )
 from alexandria3k.db_schema import ColumnMeta, TableMeta
 from alexandria3k.file_cache import get_file_cache
+from alexandria3k.json_source import JSONElementsCursor, RecordsCursor
```
We use the convention of capitalizing only the first character of acronyms.  So this would be  `JsonElementsCursor`. But see following comments.

dspinellis commented on src/alexandria3k/json_source.py:
```diff
@@ -0,0 +1,127 @@
+# pylint: disable=invalid-name
+
+
+class JSONElementsCursor(ElementsCursor):
```
I can't see how this class is related to JSON. Maybe it should be in `data_source.py` with a name such as `NestedElementsCursor`? Given the class inheritance, please think hard regarding the appropriate names of all (new and existing) classes.  Feel free to send me a diagram and discuss this.  You can also send me a simple class diagram to discuss ideas.

Please also expand the docstring to explain what this cursor expects and what functionality it provided.

dspinellis commented on src/alexandria3k/json_source.py:
```diff
@@ -0,0 +1,127 @@
+            self.elements = None
+
+
+class RecordsCursor(JSONElementsCursor):
```
The comments for  `JSONElementsCursor` also apply here, right?

dspinellis reviewed (changes requested):
Amazing work regarding the removal of code duplication. I think some restructuring is still needed. Also please squash you commits into one when resubmitting.

evgepab commented:
I moved the two cursor classes to the `data_source` module, renamed the first cursor to `NestedElementsCursor` and tried to make their functionalities clearer through their docstrings.

dspinellis reviewed (approved):
Amazing, well done! As a reminder, please send a PR to update docs/schema/datacite.dot and another one for an example (e.g. extracting metrics) under `examples/`.  I'd be happy to do the second one if you prefer.
